LibOS/shim/src/fs/chroot/fs.c

```diff
@@ -953,124 +949,135 @@ static int chroot_dput (struct shim_dentry * dent)
     if (!pal_hdl)
         return -PAL_ERRNO;
 
-    int buf_size = 0, new_size = MAX_PATH;
-    int bytes;
-    char * buf = NULL, * new_buf;
-
-    int dbufsize = MAX_PATH;
-    struct shim_dirent * dbuf = malloc(dbufsize);
-    if (!dbuf)
+    size_t buf_size = MAX_PATH, bytes = 0;
+    char * buf = malloc(buf_size);
+    if (!buf) {
+        ret = -ENOMEM;
         goto out;
-    struct shim_dirent * d = dbuf, ** last = NULL;
-
-retry:
-// ALLOCA IN A LOOP!!!
-    new_buf = __alloca(new_size);
-    if (buf)
-        memcpy(new_buf, buf, buf_size);
-    buf_size = new_size;
-    buf = new_buf;
-
-    while (1) {
-        bytes = DkStreamRead(pal_hdl, 0, buf_size, buf, NULL, 0);
+    }
 
-        if (bytes == 0) {
-            if (PAL_NATIVE_ERRNO == PAL_ERROR_ENDOFSTREAM)
-                break;
+    /*
+     * Try to read the directory list from the host. DkStramRead
+     * does not accept offset for directory listing. Therefore, we retry
+     * several times if the buffer is not large enough.
+     */
+retry_read:
+    bytes = DkStreamRead(pal_hdl, 0, buf_size, buf, NULL, 0);
+    if (!bytes) {
+        ret = 0;
+        if (PAL_NATIVE_ERRNO == PAL_ERROR_ENDOFSTREAM)
+            goto out;
 
-            if (PAL_NATIVE_ERRNO == PAL_ERROR_OVERFLOW) {
-                new_size = buf_size * 2;
-                goto retry; // FIX THIS
+        if (PAL_NATIVE_ERRNO == PAL_ERROR_OVERFLOW) {
+            char * new_buf = malloc(buf_size * 2);
+            if (!new_buf) {
+                ret = -ENOMEM;
+                goto out;
             }
 
-            ret = -PAL_ERRNO;
-            goto out;
+            free(buf);
+            buf_size *= 2;
+            buf = new_buf;
+            goto retry_read;
         }
 
-        char * b = buf, * next_b;
-        int blen;
-
-        while (b < buf + bytes) {
-            blen = strlen(b);
-            next_b = b + blen + 1;
-            bool isdir = false;
-
-            if (b[blen - 1] == '/') {
-                isdir = true;
-                b[blen - 1] = 0;
-                blen--;
-            }
+        ret = -PAL_ERRNO;
+        goto out;
+    }
 
-            int dsize = sizeof(struct shim_dirent) + blen + 1;
+    /* Now emitting the dirent data */
+    size_t dbuf_size = MAX_PATH;
+    struct shim_dirent * dbuf = malloc(dbuf_size);
+    if (!dbuf)
+        goto out;
 
-            if ((void *) d + dsize > (void *) dbuf + dbufsize) {
-                int newsize = dbufsize * 2;
-                while ((void *) d + dsize > (void *) dbuf + newsize)
-                    newsize *= 2;
+    struct shim_dirent * d = dbuf, ** last = NULL;
+    char * b = buf, * next_b;
+    int blen;
+
+    /* Scanning the directory names in the buffer */
+    while (b < buf + bytes) {
+        blen = strlen(b);
+        next_b = b + blen + 1;
+        bool isdir = false;
+
+        /* The PAL convention: if the name is ended with "/",
+           it is a directory. */
+        if (b[blen - 1] == '/') {
```

Comment:
If we assume that it's not possible to have `blen == 0` then please add an assert.